### PR TITLE
repeated headers feature with mandatory headerstart and skip

### PR DESCRIPTION
**Pull request by prabhatsir**
repeated headers feature with mandatory headerstart and skip

### CODE REVIEW

**Comment by ozlerhakan** [comment]
Hi @prabhatsir , thank you for your support :) seems like it needs to be updated with tests cases and coding

File: src/main/java/com/poiji/bind/Poiji.java [hunk 1]
```diff
@@ -359,11 +360,15 @@ private static HSSFPropertyStream deserializerPropertyStream(PoijiExcelType exce
...
      */
-    public static <T> void fromExcel(final Sheet sheet,
+    public static <T> void fromExcel(Sheet sheet,
                                      final Class<T> type,
                                      final PoijiOptions options,
                                      final Consumer<? super T> consumer) {
         Objects.requireNonNull(sheet);
+        if(options.getColumnsToKeep()!=null)
+            sheet = SheetColumns.sheetAdaptor(sheet,options);
+        options.setSkip(options.skip()-options.getHeaderStart());
```
Comment (ozlerhakan) [inline src/main/java/com/poiji/bind/Poiji.java:370]: why we need to use them here?
```
options.setSkip(options.skip()-options.getHeaderStart());
options.setHeaderStart(0);
```

File: src/main/java/com/poiji/bind/Poiji.java [hunk 2]
```diff
@@ -359,11 +360,15 @@ private static HSSFPropertyStream deserializerPropertyStream(PoijiExcelType exce
      * @see Poiji#fromExcel(Sheet, Class, PoijiOptions)
      * @see Poiji#fromExcel(Sheet, Class)
      */
-    public static <T> void fromExcel(final Sheet sheet,
+    public static <T> void fromExcel(Sheet sheet,
```
Comment (ozlerhakan) [inline src/main/java/com/poiji/bind/Poiji.java:363]: we need support all the `fromExcel` having the `Sheet` object

File: src/main/java/com/poiji/bind/mapping/SheetColumns.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+public class SheetColumns {
+
+    public static Sheet sheetAdaptor(Sheet sheet, PoijiOptions options) {
+
+        try {
+            Workbook workbook = sheet.getWorkbook();
+            FormulaEvaluator evaluator = workbook.getCreationHelper().createFormulaEvaluator();
+            evaluator.setIgnoreMissingWorkbooks(true);
+            Workbook result = new XSSFWorkbook();
+            Sheet resultSheet = result.createSheet();
```
Comment (ozlerhakan) [inline src/main/java/com/poiji/bind/mapping/SheetColumns.java:21]: creating a new sheet from scratch hmm...

File: src/main/java/com/poiji/option/PoijiOptions.java [hunk 4]
```diff
@@ -69,7 +80,7 @@ private PoijiOptions() {
         super();
     }
 
-    private PoijiOptions setSkip(int skip) {
```
Comment (ozlerhakan) [inline src/main/java/com/poiji/option/PoijiOptions.java:72]: should be private

File: src/main/java/com/poiji/option/PoijiOptions.java [hunk 5]
```diff
@@ -208,7 +219,7 @@ public int getHeaderCount() {
         return headerCount;
     }
 
-    private PoijiOptions setHeaderStart(int headerStart) {
+    public PoijiOptions setHeaderStart(int headerStart) {
```
Comment (ozlerhakan) [inline src/main/java/com/poiji/option/PoijiOptions.java:222]: should be private

File: src/main/java/com/poiji/option/PoijiOptions.java [hunk 6]
```diff
@@ -626,8 +639,18 @@ public PoijiOptionsBuilder poijiNumberFormat(final PoijiNumberFormat numberForma
...
+         * This feature needs headerstart and skip as mandatory parameters
+         * @param columnsToKeep the headers that should be included
+         */
+        public PoijiOptionsBuilder columnsToKeep(List columnsToKeep) {
+            this.columnsToKeep = columnsToKeep;
+            return this;
+        }
+
+        /**
+         * Provide columns to keep in case of repeated headers
```
Comment (ozlerhakan) [inline src/main/java/com/poiji/option/PoijiOptions.java:651]: when an excel file contains repeated columns? in which cases ?

Comment (prabhatsir) [reply src/main/java/com/poiji/option/PoijiOptions.java:651]: Sorry, I think this line creeped in from somewhere by itself

File: src/main/java/com/poiji/bind/mapping/SheetColumns.java (hunk 3 above)
Comment (prabhatsir) [reply src/main/java/com/poiji/bind/mapping/SheetColumns.java:21]: I guess better create new sheet and transfer required info to new sheet from huge old sheet.

File: src/main/java/com/poiji/option/PoijiOptions.java (hunk 4 above)
Comment (prabhatsir) [reply src/main/java/com/poiji/option/PoijiOptions.java:72]: Request support from more experienced..

File: src/main/java/com/poiji/option/PoijiOptions.java (hunk 5 above)
Comment (prabhatsir) [reply src/main/java/com/poiji/option/PoijiOptions.java:222]: Request support from more experienced..

File: src/main/java/com/poiji/bind/Poiji.java (hunk 2 above)
Comment (prabhatsir) [reply src/main/java/com/poiji/bind/Poiji.java:363]: Request support from more experienced..

File: src/main/java/com/poiji/bind/Poiji.java (hunk 1 above)
Comment (prabhatsir) [reply src/main/java/com/poiji/bind/Poiji.java:370]: skip was itself increasing itself to skip+headerstart.
in new virtual sheet the headers are at 0
Request support from more experienced..

**Comment by ozlerhakan** [comment]
Hi @prabhatsir , when do you want to update the corresponding parts?

**Comment by prabhatsir** [comment]
@ozlerhakan when time allows

**Comment by ozlerhakan** [comment]
Hi @prabhatsir , it has been 4 months, I'm closing it but feel free to make your changes, we can re-open it :)

**Pull request closed without merging**